### PR TITLE
3599 api for simple dataset edits

### PR DESCRIPTION
**Pull request by sekmiller**
## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Thanks!

## Related Issues

- connects to #3599 Add API for Dataset edits

## Pull Request Checklist

- [x] Unit [tests][] completed
- [x] Integration [tests][]: Completed
- [x] Deployment requirements: [x] None
- [x] [Documentation][docs] completed
- [x] Merged latest from "develop" [branch][x] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: https://github.com/IQSS/dataverse/tree/develop/scripts/database/upgrades
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/DatasetField.java [hunk 1]
```diff
@@ -420,6 +420,10 @@ public Dataverse getDataverse() {
         } else if (template != null) {
             return template.getDataverse();
         } else {
+
+            System.out.print("getDataverseException: " + this.datasetFieldType.getDisplayName());
```
Comment (matthew-a-dunlap) [inline src/main/java/edu/harvard/iq/dataverse/DatasetField.java]: Should we use logger instead?

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java [hunk 2]
```diff
@@ -407,6 +413,105 @@ public Response updateDraftVersion( String jsonBody, @PathParam("id") String id,
...
+    public Response addDataToVersion(String jsonBody, @PathParam("id") String id) throws WrappedResponse{
+
+        DataverseRequest req = createDataverseRequest(findUserOrDie());
+
+        return processDatasetUpdate(jsonBody, id, req, true);
+    }
+    
+    @PUT
+    @Path("{id}/updateData")
+    public Response updateDataInVersion(String jsonBody, @PathParam("id") String id)  throws WrappedResponse {
```
Comment (matthew-a-dunlap) [inline src/main/java/edu/harvard/iq/dataverse/api/Datasets.java]: Both endpoints do the exact same thing? A usability decision?

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java [hunk 3]
```diff
@@ -407,6 +413,105 @@ public Response updateDraftVersion( String jsonBody, @PathParam("id") String id,
...
+                singleField  = jsonParser().parseField(json, Boolean.FALSE);
+                fields.add(singleField);
+            } else{
+                fields = jsonParser().parseMultipleFields(json);
+            }
+
+            dsv.setVersionState(DatasetVersion.VersionState.DRAFT);
+                
+
+            for (DatasetField dsf : dsv.getDatasetFields()) {
```
Comment (matthew-a-dunlap) [inline src/main/java/edu/harvard/iq/dataverse/api/Datasets.java]: A comment about this parser-loop would probably help future maintenance.

File: src/test/java/edu/harvard/iq/dataverse/api/DatasetsIT.java [hunk 4]
```diff
@@ -128,6 +128,78 @@ public void testCreateDataset() {
...
+        Response deleteTitleViaNative = UtilIT.updateFieldLevelDatasetMetadataViaNative(datasetPersistentId, pathToJsonFileBadData, apiToken);
+        deleteTitleViaNative.prettyPrint();
+        deleteTitleViaNative.then().assertThat().body("message", equalTo("Validation failed: Title is required."));
+
+        
+        Response publishDataverse = UtilIT.publishDataverseViaSword(dataverseAlias, apiToken);
+        
+        Response publishDataset = UtilIT.publishDatasetViaNativeApi(datasetPersistentId, "major", apiToken);
+        assertEquals(200, publishDataset.getStatusCode());
+        //post publsih update
```
Comment (matthew-a-dunlap) [inline src/test/java/edu/harvard/iq/dataverse/api/DatasetsIT.java]: typo

File: src/main/java/edu/harvard/iq/dataverse/DatasetField.java (hunk 1 above)
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/DatasetField.java]: Yes.

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java [hunk 5]
```diff
@@ -407,6 +413,105 @@ public Response updateDraftVersion( String jsonBody, @PathParam("id") String id,
         }
     }
     
+    @PUT
+    @Path("{id}/addData")
+    public Response addDataToVersion(String jsonBody, @PathParam("id") String id) throws WrappedResponse{
+
+        DataverseRequest req = createDataverseRequest(findUserOrDie());
```
Comment (matthew-a-dunlap) [inline src/main/java/edu/harvard/iq/dataverse/api/Datasets.java]: From discussion with @sekmiller this boolean should be false

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java (hunk 3 above)
Comment (sekmiller) [reply src/main/java/edu/harvard/iq/dataverse/api/Datasets.java]: Added a comment and also changed the name of the looping variable to add clarity

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java (hunk 2 above)
Comment (sekmiller) [reply src/main/java/edu/harvard/iq/dataverse/api/Datasets.java]: mistake as noted below

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java (hunk 5 above)
Comment (sekmiller) [reply src/main/java/edu/harvard/iq/dataverse/api/Datasets.java]: fixed

File: src/test/java/edu/harvard/iq/dataverse/api/DatasetsIT.java (hunk 4 above)
Comment (sekmiller) [reply src/test/java/edu/harvard/iq/dataverse/api/DatasetsIT.java]: fixed

File: src/main/java/edu/harvard/iq/dataverse/DatasetField.java (hunk 1 above)
Comment (sekmiller) [reply src/main/java/edu/harvard/iq/dataverse/DatasetField.java]: Really don't need to log it. It's debug code that I should have removed. Throwing the exception is enough, when this state legitimately occurs.

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java [hunk 6]
```diff
@@ -407,6 +413,106 @@ public Response updateDraftVersion( String jsonBody, @PathParam("id") String id,
...
+                
+            //loop through the existing dataset fields in the draft version
+            // and compare to the update fields
+            for (DatasetField dsf : dsv.getDatasetFields()) {
+                for (DatasetField updateField : fields) {
+                    if (dsf.getDatasetFieldType().equals(updateField.getDatasetFieldType())) {
+                        if (dsf.isEmpty() || dsf.getDatasetFieldType().isAllowMultiples() || replaceData ) {
+                            if(replaceData){
+                                if(dsf.getDatasetFieldType().isAllowMultiples()){
+                                    dsf.setDatasetFieldCompoundValues(new ArrayList());
```
Comment (oscardssmith) [inline src/main/java/edu/harvard/iq/dataverse/api/Datasets.java]: Should use `ArrayList<>()` for type stability

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java [hunk 7]
```diff
@@ -407,6 +413,106 @@ public Response updateDraftVersion( String jsonBody, @PathParam("id") String id,
...
+                fields.add(singleField);
+            } else{
+                fields = jsonParser().parseMultipleFields(json);
+            }
+
+            dsv.setVersionState(DatasetVersion.VersionState.DRAFT);
+                
+            //loop through the existing dataset fields in the draft version
+            // and compare to the update fields
+            for (DatasetField dsf : dsv.getDatasetFields()) {
```
Comment (oscardssmith) [inline src/main/java/edu/harvard/iq/dataverse/api/Datasets.java]: I don't suppose there is a way to make this less nested?

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java [hunk 8]
```diff
@@ -526,6 +534,23 @@ private Response processDatasetUpdate(String jsonBody, String id, DataverseReque
         }
     }
     
+    private void validateDatasetFieldValues(List<DatasetField> fields) throws JsonParseException {
+        String errors = "";
+
+        for (DatasetField dsf : fields) {
+            if (dsf.getDatasetFieldType().isAllowMultiples() && dsf.getControlledVocabularyValues().isEmpty()
+                    && dsf.getDatasetFieldCompoundValues().isEmpty() && dsf.getDatasetFieldValues().isEmpty()) {
+                errors += "Empty multiple value for field " + dsf.getDatasetFieldType().getDisplayName() + " ";
```
Comment (oscardssmith) [inline src/main/java/edu/harvard/iq/dataverse/api/Datasets.java]: should use `StringBuilder` to avoid bad perf in the case of many errors.

Comment (sekmiller) [reply src/main/java/edu/harvard/iq/dataverse/api/Datasets.java]: OK. Done. thanks, Oscar.

**Pull request merged**